### PR TITLE
improve(common): convert common to typescript

### PR DESCRIPTION
**Pull request by mrice32**
Motivation**

We'd like all of our packages to be written in typescript.

**Summary**

This PR converts common to typescript. It _mostly_ does so faithfully with as few `any`s as possible.

Note: some files and functions that are unused were removed rather than converted.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [x]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

N/A

### CODE REVIEW

**Review by daywiss — APPROVED** [review]
this looks really good, and great job typing almost everything on this massive pr

File: packages/common/src/HardhatConfig.ts [hunk 1]
```diff
@@ -88,18 +93,20 @@ function getHardhatConfig(configOverrides, workingDir = "./", includeTruffle = t
       apiKey: process.env.ETHERSCAN_API_KEY,
     },
     namedAccounts: { deployer: 0 },
-  };
+  } as unknown) as HardhatConfig; // Cast to allow extra properties.
```
[the remark this answers is not in the document]
Comment (mrice32) [reply packages/common/src/HardhatConfig.ts:96]: Typescript makes you do that if "there isn't sufficient overlap" between the types.

File: packages/common/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import type _Web3 from "web3";
+
+export type BN = ReturnType<_Web3["utils"]["toBN"]>;
```
[the remark this answers is not in the document]
Comment (mrice32) [reply packages/common/src/types.ts:3]: Yeah :/

File: packages/common/src/ObjectUtils.ts [hunk 3]
```diff
@@ -13,7 +13,7 @@
  * the new Object and `isValid` will be called to validate each of `newObject`'s properties.
  * @return `newObject` a new Object with the same properties as `defaultProps`, or `defaultProps` if undefined `overrideProps`.
  */
-const createObjectFromDefaultProps = (overrideProps, defaultProps) => {
+export const createObjectFromDefaultProps = (overrideProps: any, defaultProps: any): any => {
```
[the remark this answers is not in the document]
Comment (mrice32) [reply packages/common/src/ObjectUtils.ts]: I think they're just general object, so I used your first suggestion!

**Pull request merged**